keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/optimizer/KeanuProbabilisticGraph.java

```diff
@@ -0,0 +1,115 @@
+
+    private final Map<String, Vertex> vertexLookup;
+
+    @Getter
+    private final List<Vertex> latentVertices;
+
+    @Getter
+    private final List<Vertex> observedVertices;
+
+    @Getter
+    private final List<Vertex> latentOrObservedVertices;
+
+    public KeanuProbabilisticGraph(BayesianNetwork bayesianNetwork) {
+
+        this.vertexLookup = bayesianNetwork.getLatentVertices().stream()
+            .collect(toMap(KeanuProbabilisticGraph::getUniqueStringReference, v -> v));
+
+        this.latentVertices = ImmutableList.copyOf(bayesianNetwork.getLatentVertices());
+        this.observedVertices = ImmutableList.copyOf(bayesianNetwork.getObservedVertices());
+        this.latentOrObservedVertices = ImmutableList.copyOf(bayesianNetwork.getLatentOrObservedVertices());
+    }
+
+    @Override
+    public double logProb(Map<String, ?> inputs) {
+        cascadeUpdate(inputs);
+        return ProbabilityCalculator.calculateLogProbFor(this.latentOrObservedVertices);
+    }
+
+    @Override
+    public double logLikelihood(Map<String, ?> inputs) {
+        cascadeUpdate(inputs);
+        return ProbabilityCalculator.calculateLogProbFor(this.observedVertices);
+    }
+
+    @Override
+    public List<String> getLatentVariables() {
+        return ImmutableList.copyOf(
+            this.latentVertices.stream()
+                .map(KeanuProbabilisticGraph::getUniqueStringReference)
+                .collect(Collectors.toList())
+        );
+    }
+
+    @Override
+    public Map<String, ?> getLatentVariablesValues() {
+        return latentVertices.stream()
+            .collect(toMap(
+                KeanuProbabilisticGraph::getUniqueStringReference,
+                Vertex::getValue)
+            );
+    }
+
+    @Override
+    public Map<String, long[]> getLatentVariablesShapes() {
+        Map<String, ?> latentVariablesValues = getLatentVariablesValues();
+        return getLatentVariables().stream()
+            .collect(Collectors.toMap(
+                v -> v,
+                v -> ((Tensor) latentVariablesValues.get(v)).getShape())
+            );
+    }
+
+    public void cascadeUpdate(Map<String, ?> inputs) {
+
+        List<Vertex> updatedVertices = new ArrayList<>();
+        for (Map.Entry<String, ?> input : inputs.entrySet()) {
+            Vertex updatingVertex = vertexLookup.get(input.getKey());
+
+            if (updatingVertex == null) {
+                throw new IllegalArgumentException("Cannot cascade update for input: " + input.getKey());
+            }
+
+            updatingVertex.setValue(input.getValue());
+            updatedVertices.add(updatingVertex);
+        }
+
+        VertexValuePropagation.cascadeUpdate(updatedVertices);
+    }
+
+    /**
+     * @param vertex to get unique string for
+     * @return A string that is unique to this vertex within any graph that this
+     * vertex is a member of.
+     */
+    public static String getUniqueStringReference(Vertex vertex) {
```

Comment:
I don't like the idea of using Strings everywhere. Could we define a new interface `VariableReference` instead, and have `VertexLabel` and `VertexId` implement it? It would probably just need one method `String getValue` but it goes a long way to keeping the code base type safe.

Reply:
also we could define an interface that `Vertex` implements with the methods that you use in this PR:
`VariableReference getReference()`
`Object getValue()`
`long[] getShape`

Reply:
I like the idea of creating a VariableReference interface as part of the ProbabilisticGraph interfaces. It cleans up the VariableBounds class as well.